NIFI-2292: Funnel all cluster node status changes through the cluster coordinator…

markap14 opened the pull request:
… instead of having each node broadcast changes to the whole cluster. This gives us the ability to increment the updateId consistently without race conditions.

JPercivall commented:
Ocasionally see error:
3 node secure cluster
have non-coordinator node disconnect itself
from the other non-node, refresh cluster panel
See "An unexpected error has occurred" in panel

![screen shot 2016-07-22 at 2 42 48 pm](https://cloud.githubusercontent.com/assets/11302527/17068754/d6883360-5020-11e6-8a84-e9d879426dbd.png)

JPercivall commented:
I think that error is reproducible when I disconnect the primary node.

JPercivall commented:
@markap14, this branch has conflicts with the current master, can you resolve them?

JPercivall commented:
2 reproducible errors:

3 node secure cluster
node 1 is primary and coordinator
view bulletin board from node 2
using node 1 disconnect node 3
first error: see multiple messages in node 2 regarding this event 
using node 1 disconnect self
Second error: in node 2 bulletin board see continuous stream of error messages

![screen shot 2016-07-25 at 10 13 43 am](https://cloud.githubusercontent.com/assets/11302527/17104423/d7818992-5250-11e6-8d49-0bb4bce2c4af.png)
![screen shot 2016-07-25 at 10 10 39 am](https://cloud.githubusercontent.com/assets/11302527/17104428/d9960334-5250-11e6-90ea-fad8c552fd47.png)

markap14 commented:
@JPercivall that's a good catch! I updated the PR to address this. The issue was that whenever there is only 1 node in the cluster, it is complaining if we attempt to disconnect ANY node (even one that's already marked disconnected). And since we'd received a heartbeat from a disconnected node, we kept trying to disconnect it. Updated code and unit tests to verify. Thanks!

markap14 commented:
OK So I addressed your comments above, @JPercivall. I then moved on and reviewed PR 702 and made a typo when I went to merge PR 702. I accidentally merged PR 701 (this one) and pushed to master. I then reverted that. When I then rebased my local branch against master, I had some problems.

To make sure that all is in a consistent state, I created a new branch (NIFI-2292-rebased) and then created a new PR from that branch (PR 717). Please finish reviewing based off of that PR, not this one. Sorry about that!
